File: 0x10 基本数据结构/习题/AcWing151.cpp
#include <iostream>
#include <stack>
#include <cctype>
using namespace std;
stack<int> nums;
stack<char> ops;

void calc() {
  int b = nums.top(); nums.pop();
  int a = nums.top(); nums.pop();
  char c = ops.top(); ops.pop();
  int d = 1;
  switch (c) {
  case '+': d = a + b; break;
  case '-': d = a - b; break;
  case '*': d = a * b; break;
  case '/': d = a / b; break;
  case '^': while (b--) d *= a;
  }
  nums.push(d);
}

int main() {
  string s;
  cin >> s;
  s = '(' + s + ')'; // 首尾加括号，强制算到最后栈中只有一个数
  int l = 0, r = 0; // 计算左右括号数，若个数不匹配需补到相等
  for (int i=0; i<s.size(); i++) {
    if (s[i] == '(') l++;
    if (s[i] == ')') r++;
  }
  if (l >= r) s.append(l - r, ')');
  else s = string(r - l, '(') + s;
  for (int i=0; i<s.size(); i++) {
    if (isdigit(s[i])) {
      int j = i, t = 0;
      while (isdigit(s[j])) t = t*10 + s[j++]-'0';
      i = j - 1;
      nums.push(t);
    } else {
      char c = s[i];
      if (c == '(') ops.push(c);
      else if (c == ')') {
        while (ops.top() != '(') calc();
        ops.pop();
      } else if (c == '+' || c == '-') {
        // 如果当前这个'-'代表负号而非减号
        if (c=='-' && i && !isdigit(s[i-1]) && s[i-1]!=')') {
          if (s[i+1] == '(')
            nums.push(-1), ops.push('*');
          else {
            int j = i+1, t = 0;
            while (isdigit(s[j])) t = t*10 + s[j++]-'0';
            i = j - 1;
            nums.push(-t);
          }
        } else { // +-优先级最低，若栈中有其他运算符，直接计算
          while (ops.top() != '(') calc();
          ops.push(c);
        }
      } else if (c == '*' || c == '/') {
        while (ops.top()=='*' || ops.top()=='/' || ops.top()=='^') calc();
        ops.push(c);
      } else if (c == '^') {
        while (ops.top() == '^') calc();
        ops.push(c);
      }
    }
  }
  cout << nums.top() << endl;
  return 0;
}

File: 0x10 基本数据结构/习题/AcWing151.py
nums, ops = [], []

def calc():
    b = nums.pop()
    a = nums.pop()
    c = ops.pop()
    if c == '+': d = a + b
    elif c == '-': d = a - b
    elif c == '*': d = a * b
    elif c == '/': d = int(a / b)
    else: d = a**b
    nums.append(d)
    
def main():
    s = input()
    s = '(' + s + ')'
    l, r = 0, 0
    for i in range(len(s)):
        if s[i] == '(': l += 1
        if s[i] == ')': r += 1
    if l >= r: s = s + ')'*(l-r)
    else: s = '('*(r-l) + s
    i = 0
    while i < len(s):
        if s[i].isdigit():
            j, t = i, 0
            while s[j].isdigit(): t = t*10 + ord(s[j])-ord('0'); j += 1
            i = j-1
            nums.append(t)
        else:
            c = s[i]
            if c == '(': ops.append(c)
            elif c == ')':
                while ops[-1] != '(': calc()
                ops.pop()
            elif c == '+' or c == '-':
                if c=='-' and i and not s[i-1].isdigit() and s[i-1]!=')':
                    if s[i+1] == '(':
                        nums.append(-1); ops.append('*')
                    else:
                        j, t = i+1, 0
                        while s[j].isdigit(): t = t*10 + ord(s[j])-ord('0'); j += 1
                        i = j-1
                        nums.append(-t)
                else:
                    while ops[-1] != '(': calc()
                    ops.append(c)
            elif c == '*' or c == '/':
                while ops[-1]=='*' or ops[-1]=='/' or ops[-1]=='^': calc()
                ops.append(c)
            elif c == '^':
                while ops[-1] == '^': calc()
                ops.append(c)
        i += 1
    print(nums[-1])
    
main()